packages/admin-ui/src/segment/segment.tsx

```diff
@@ -0,0 +1,255 @@
+  const hourFormatter = useDateFormatter({
+    hour: 'numeric',
+    hour12: dateFormatter.resolvedOptions().hour12,
+  })
+
+  const { spinButtonProps } = useSpinButton({
+    value: segment.value,
+    textValue: getTextValue(segment, state, {
+      month: monthFormatter,
+      hour: hourFormatter,
+    }),
+    minValue: segment.minValue,
+    maxValue: segment.maxValue,
+    isDisabled,
+    isReadOnly,
+    isRequired,
+    onIncrement: () => increment(segment.type),
+    onDecrement: () => decrement(segment.type),
+    onIncrementPage: () => incrementPage(segment.type),
+    onDecrementPage: () => decrementPage(segment.type),
+    onIncrementToMax: () =>
+      setSegment(segment.type, segment.maxValue as number),
+    onDecrementToMin: () =>
+      setSegment(segment.type, segment.minValue as number),
+  })
+
+  const onNumericKeyDown = useCallback(
+    (key: string) => {
+      const newValue = enteredKeys.concat(key)
+
+      if (segment.type === 'dayPeriod') {
+        if (key === 'a') {
+          state.setSegment('dayPeriod', 0)
+        } else if (key === 'p') {
+          state.setSegment('dayPeriod', 12)
+        }
+
+        state.next()
+      } else {
+        if (!isNumeric(newValue)) {
+          return
+        }
+
+        const numberValue = parseNumber(newValue)
+        let segmentValue = numberValue
+
+        if (
+          segment.type === 'hour' &&
+          state.dateFormatter.resolvedOptions().hour12 &&
+          numberValue === 12
+        ) {
+          segmentValue = 0
+        } else if (numberValue > (segment.maxValue as number)) {
+          segmentValue = parseNumber(key)
+        }
+
+        state.setSegment(segment.type, segmentValue)
+
+        if (Number(`${numberValue}0`) > (segment.maxValue as number)) {
+          setEnteredKeys('')
+          state.next()
+        } else {
+          setEnteredKeys(newValue)
+        }
+      }
+    },
+    [enteredKeys, next, segment.maxValue, segment.type]
+  )
+
+  const onKeyDown = useCallback(
+    (e: React.KeyboardEvent) => {
+      if (e.ctrlKey || e.metaKey || e.shiftKey || e.altKey) {
+        return
+      }
+
+      switch (e.key) {
+        case 'Enter':
+          e.preventDefault()
+          next()
+          break
+
+        case 'Tab':
+          break
+
+        case 'Backspace': {
+          e.preventDefault()
+          if (isNumeric(segment.text) && !isReadOnly) {
+            const newValue = segment.text.slice(0, -1)
+
+            setSegment(
+              segment.type,
+              newValue.length === 0
+                ? (segment.minValue as number)
+                : parseNumber(newValue)
+            )
+            setEnteredKeys(newValue)
+          }
+
+          break
+        }
+
+        default:
+          e.preventDefault()
+          e.stopPropagation()
+          if ((isNumeric(e.key) || /^[ap]$/.test(e.key)) && !isReadOnly) {
+            // TODO: fix typing experience
```

Comment:
Worth creating an issue?

Reply:
I don't think so, this is part of an experimental component and we'll revisit it anyways in the future